launcher_test.go

```diff
@@ -33,45 +34,97 @@ func (t *testWorker) DoTask(ctx context.Context) (Result, string) {
 
 func TestLauncher_Status(t *testing.T) {
 	//setup
-	prevMsg := nop.FakeMsg
-	defer func() {
-		nop.FakeMsg = prevMsg
-	}()
-	c, _ := nop.NewConsumer("")
+	c, _ := nop.NewConsumer(nop.Repeat, `{"type":"test","info":"?wait-time=5s"}`)
 	p, _ := nop.NewProducer("")
 
 	// Test MaxInProgress is 100
-	nop.FakeMsg = []byte(`{"type":"test","info":"?wait-time=5s"}`)
 	l := NewLauncherFromBus(newTestWorker, c, p, &LauncherOptions{MaxInProgress: 100})
-	l.DoTasks()
+	_, cancel := l.DoTasks()
+
 	time.Sleep(100 * time.Millisecond)
 	sts := l.Stats()
+	cancel()
 	if sts.TasksRunning != 100 {
 		t.Errorf("Tasks running should be 100 !=%d", sts.TasksRunning)
+	} else {
+		t.Log("PASS: 100 concurrent workers")
 	}
 
 	// Test Average Run time - single worker
-	nop.FakeMsg = []byte(`{"type":"test","info":"?wait-time=10ms"}`)
+	c, _ = nop.NewConsumer("", `{"type":"test","info":"?wait-time=10ms"}`)
 	l = NewLauncherFromBus(newTestWorker, c, p, nil)
 	l.DoTasks()
 	time.Sleep(100 * time.Millisecond)
 	sts = l.Stats()
 	if sts.MeanTaskTime != "10ms" {
 		t.Errorf("Status should display MeanTaskTime: 10ms %v", sts.MeanTaskTime)
+	} else {
+		t.Log("PASS: 10ms mean")
 	}
 
 	// Test Average Run time - multiple workers
-	nop.FakeMsg = []byte(`{"type":"test","info":"?wait-time=10ms"}`)
+	c, _ = nop.NewConsumer(nop.Repeat, `{"type":"test","info":"?wait-time=10ms"}`)
 	l = NewLauncherFromBus(newTestWorker, c, p, &LauncherOptions{MaxInProgress: 20})
 	l.DoTasks()
 	time.Sleep(100 * time.Millisecond)
 	sts = l.Stats()
 	if sts.MeanTaskTime != "10ms" {
 		t.Errorf("Status should display MeanTaskTime: 10ms %v", sts.MeanTaskTime)
+	} else {
+		t.Log("PASS: 20 concurrent workers with 10ms mean")
 	}
 
 }
 
+// TestLauncher_Limiter verifies the rate limiter features.
+// The first time a task comes in tasks up to the MaxInProgress
+// should immediately start and then will be throttled by the hourly rate limit
+func TestLauncher_Limiter(t *testing.T) {
+	const RateLimit = 36000 // or 1 task every 100ms
+
+	t.Run("1 concurrent worker with 100ms rate limit", func(t *testing.T) {
+
+		//setup
+		c, _ := nop.NewConsumer(nop.Repeat, `{"type":"test","info":"?wait-time=10ms"}`)
+		p, _ := nop.NewProducer("")
+
+		l := NewLauncherFromBus(newTestWorker, c, p, &LauncherOptions{MaxInProgress: 1, TaskLimit: RateLimit})
+		_, cancel := l.DoTasks()
+		time.Sleep(25 * time.Millisecond)
+		sts := l.Stats()
+		if !(sts.TasksRunning == 0 && sts.Producer.Sent["done"] == 1) {
+			t.Fatal(spew.Sdump(sts))
+		}
+
+		time.Sleep(200 * time.Millisecond)
+		if !(sts.TasksRunning == 0 && sts.Producer.Sent["done"] == 3 && sts.MeanTaskTime == "10ms") {
+			t.Fatal(spew.Sdump(sts))
+		}
+		cancel()
+	})
+
+	t.Run("5 concurrent workers with 100ms rate limit", func(t *testing.T) {
+		c, _ := nop.NewConsumer(nop.Repeat, `{"type":"test","info":"?wait-time=10ms"}`)
```

Comment:
shouldn't the comment be 10ms not 100ms?

Reply:
The worker takes 10ms to run, but the rate limit is set to 100ms. So we'd expected a worker to run every 100ms, but for its average task time to be 10ms.